### PR TITLE
SQL: add secondary indexes to INSERT/DELETE

### LINKED ISSUE #1789: sql: support secondary indexes in INSERT
`INSERT` currently only creates key/value pairs the for the columns. We also need to create key/value pairs for any secondary indexes.

### PR DESCRIPTION
**Pull request by tamird**
@cockroachdb/structured 

Closes #1789.

### CODE REVIEW

File: sql/insert.go [hunk 1]
```diff
@@ -92,25 +111,27 @@ func (p *planner) Insert(n *parser.Insert) (planNode, error) {
 	if err := rows.Err(); err != nil {
 		return nil, err
 	}
-	if err := p.db.Run(&b); err != nil {
+	if err := p.db.Txn(func(txn *client.Txn) error {
```
Comment (vivekmenezes) [inline sql/insert.go]: Why did you add a transaction here? We want to implement transactions at a higher layer to include many inserts, etc

File: sql/delete.go [hunk 2]
```diff
@@ -63,17 +63,36 @@ func (p *planner) Delete(n *parser.Delete) (planNode, error) {
...
+
+		// Delete the secondary indexes.
+		for _, secondaryIndex := range tableDesc.Indexes[1:] {
+			secondaryIndexKeyPrefix := encodeIndexKeyPrefix(tableDesc.ID, secondaryIndex.ID)
+			secondaryIndexKey, err := encodeIndexKey(secondaryIndex, colMap, values, secondaryIndexKeyPrefix)
+			if err != nil {
+				return nil, err
+			}
+			rowStartKey := proto.Key(append(secondaryIndexKey, primaryIndexKeySuffix...))
+			b.DelRange(rowStartKey, rowStartKey.PrefixEnd())
```
Comment (petermattis) [inline sql/delete.go]: For secondary indexes, each index will correspond to a single key. It would be more efficient to just use `Del(rowStartKey)` here.

Comment (tamird) [inline sql/delete.go]: Yeah, just made that change.

File: sql/insert.go [hunk 3]
```diff
@@ -60,19 +60,38 @@ func (p *planner) Insert(n *parser.Insert) (planNode, error) {
...
+		// Write the secondary indexes.
+		for _, secondaryIndex := range tableDesc.Indexes[1:] {
+			secondaryIndexKeyPrefix := encodeIndexKeyPrefix(tableDesc.ID, secondaryIndex.ID)
+			secondaryIndexKey, err := encodeIndexKey(secondaryIndex, colMap, values, secondaryIndexKeyPrefix)
+			if err != nil {
+				return nil, err
+			}
+
+			// TODO:(tamird): check uniqueness constraints (primary and
+			// secondary). Probably will have to do some scans?
```
Comment (vivekmenezes) [inline sql/insert.go]: I think you can factor out the secondary index key construction into a method here in insert or in keys/... and add some unittests for the key construction. That way the code in here is just inserting/deleting keys.

**Comment by tamird** [comment]
@petermattis can you help me understand why the second commit here fixes the tests? I'm apparently not 100% on `append`'s magic.

**Comment by tschottdorf** [comment]
the LHS appends to the contiguous memory owned by `primaryIndexKeyPrefix` (possibly re-allocating). RHS creates a copy. At the very least it's a data race.

**Comment by vivekmenezes** [comment]
https://news.ycombinator.com/item?id=9555472

**Comment by tamird** [comment]
PTAL

**Comment by vivekmenezes** [comment]
LGTM

File: sql/insert.go [hunk 4]
```diff
@@ -88,6 +95,22 @@ func (p *planner) Insert(n *parser.Insert) (planNode, error) {
 				b.Put(key, string(t))
```
Comment (petermattis) [inline sql/insert.go]: I think if you used `CPut` instead of `Put` here and required no existing value to be present (`expValue == nil`) then we would get the appropriate error on trying to insert an already existing value. Same idea could be used for other unique indexes.

File: sql/delete.go [hunk 5]
```diff
@@ -63,14 +66,35 @@ func (p *planner) Delete(n *parser.Delete) (planNode, error) {
...
+
+		// Delete the secondary indexes.
+		secondaryIndexKeys, err := encodeSecondaryIndexKeys(tableDesc, colMap, values, primaryIndexKeySuffix)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, secondaryIndexKey := range secondaryIndexKeys {
+			// TODO:(tamird): check uniqueness constraints (primary and
+			// secondary). Probably will have to do some scans?
```
Comment (petermattis) [inline sql/delete.go]: Why do you have to check uniqueness constraints on delete? I would think that is only necessary on INSERT and UPDATE.

File: sql/insert.go (hunk 4 above)
Comment (tamird) [inline sql/insert.go]: This would work for the primary key, but not for secondary keys, since the uniqueness constraint applies to a prefix of the proper key.

File: sql/delete.go (hunk 5 above)
Comment (tamird) [inline sql/delete.go]: blind c&p. removed.

**Comment by petermattis** [comment]
No tests? I think we should have some tests in the `sql` directory that inserting/deleting affecting the expected keys.

File: sql/insert.go (hunk 4 above)
Comment (petermattis) [inline sql/insert.go]: Non-unique secondary keys are encoded as `<secondary-key>/<primary-key> -> NULL`. Perhaps unique secondary keys should be encoded as `<secondary-key> -> <primary-key>`. This makes unique for non-unique indexes differ, but allows the continued use of `CPut` which is a pretty big advantage.

**Comment by petermattis** [comment]
LGTM

File: sql/insert.go (hunk 4 above)
Comment (tamird) [inline sql/insert.go]: That's a good idea. Since the main disadvantage of this approach comes on the querying side, I'll let you make the decision in your select-on-secondary index work.

Comment (tamird) [inline sql/insert.go]: I've changed the primary key case to a CPut.

**Comment by tamird** [comment]
> No tests? I think we should have some tests in the sql directory that inserting/deleting affecting the expected keys.

Yeah, I wasn't sure about this. We seem to have gone fully the integration test route so far, so I hesitated to deviate. How were you planning on testing secondary indexes on the select side?

File: sql/insert.go (hunk 4 above)
Comment (petermattis) [inline sql/insert.go]: I think the benefit on INSERT/UPDATE outweighs the added complexity to SELECT. Mind creating an issue for this if you don't plan on doing it immediately?

**Comment by petermattis** [comment]
I think the vast majority of the tests should be verifications of SQL
statements and results. But there should be a handful of tests that verify
that simple SQL data modifications are affecting the expected keys and
values.

Testing of the SELECT support for indexes will likely be a mixture of
SQL-level tests and lower-level unit tests that verify indexes are being
selected given appropriate table definitions and WHERE clauses.

File: sql/insert.go [hunk 6]
```diff
@@ -60,33 +61,55 @@ func (p *planner) Insert(n *parser.Insert) (planNode, error) {
...
 			case parser.DInt:
-				b.Put(key, int64(t))
+				b.CPut(key, int64(t), nil)
 			case parser.DFloat:
-				b.Put(key, float64(t))
+				b.CPut(key, float64(t), nil)
 			case parser.DString:
-				b.Put(key, string(t))
+				b.CPut(key, string(t), nil)
+			}
```
Comment (petermattis) [inline sql/insert.go]: See my comment in @vivekmenezes's change. I think we should structure this as `b.CPut(key, prepareVal(val), nil)`. Fine to do so in a separate PR.

**Comment by tamird** [comment]
@petermattis I've implemented uniqueness constraints and updated the design doc accordingly. PTAL.

File: docs/structured.md [hunk 7]
```diff
@@ -150,15 +150,19 @@ changing the primary key.
...
-can be enumerated under a single SecondaryKey. A lookup will involve looking up
-the secondary index using the secondary key to pick up all the primary keys, and
-further using the primary keys to get to all the data rows. A row insertion
-involves computing the primary and secondary keys for the row, adding the data
-under the primary key, and adding the primary key to all the secondary indexes
-under the secondary keys. A row deletion behaves analogously.
+keys. Unique secondary keys will map directly to the primary key via the
+following anatomy: `/TableID/SecondaryIndexID/SecondaryKey` -> `PrimaryKey`.
+Non-unique secondary keys must be able to address multiple primary keys, and
+will so will use a prefix-based anatomy:
```
Comment (petermattis) [inline docs/structured.md]: `s/will so will/so will/g`

File: sql/driver/driver_test.go [hunk 8]
```diff
@@ -350,6 +352,19 @@ func TestInsertSelectDelete(t *testing.T) {
 		if _, err := db.Exec(fmt.Sprintf(`INSERT INTO %s.kv VALUES ('e', 'f')`, testcase.db)); err != nil {
 			t.Fatal(err)
 		}
+		// Primary key uniqueness violation.
+		if _, err := db.Exec(fmt.Sprintf(`INSERT INTO %s.kv VALUES ('e', 'f')`, testcase.db)); !isError(err, "unexpected value") {
```
Comment (petermattis) [inline sql/driver/driver_test.go]: We might want to translate the `CPut` error into something more informative.

**Comment by petermattis** [comment]
LGTM

File: docs/structured.md (hunk 7 above)
Comment (tamird) [inline docs/structured.md]: done

File: sql/driver/driver_test.go (hunk 8 above)
Comment (tamird) [inline sql/driver/driver_test.go]: done. left a TODO to figure out how to print the specific index that is being violated.

File: sql/insert.go [hunk 9]
```diff
@@ -93,24 +115,27 @@ func (p *planner) Insert(n *parser.Insert) (planNode, error) {
 		return nil, err
 	}
 	if err := p.db.Run(&b); err != nil {
+		if tErr, ok := err.(*proto.ConditionFailedError); ok {
+			return nil, fmt.Errorf("duplicate key value %q violates unique constraint %s", tErr.ActualValue.Bytes, "TODO(tamird)")
+		}
```
Comment (petermattis) [inline sql/insert.go]: The error returned by `p.db.Run(&b)` is the first error encountered in the batch. I think you can loop over `b.Results` to find the first one that returned an error and match it up with the index being modified. Irritating, but likely worthwhile. 

You definitely need to be running the batch within a transaction. Could have sworn that was being done already.

Comment (tamird) [inline sql/insert.go]: Discussed this with @vivekmenezes today - we can't just instantiate a transaction here since this statement can execute as part of a SQL transaction. In other words, the transaction has to be passed into this method.

I filed https://github.com/cockroachdb/cockroach/issues/1889 earlier today to track.

Comment (tamird) [inline sql/insert.go]: Looks like there's a bug(?) in batch that prevents looping over `b.Results`. See #1891.

File: sql/insert.go [hunk 10]
```diff
@@ -60,32 +62,52 @@ func (p *planner) Insert(n *parser.Insert) (planNode, error) {
...
+				b.CPut(key, bool(t), nil)
 			case parser.DInt:
-				b.Put(key, int64(t))
+				b.CPut(key, int64(t), nil)
 			case parser.DFloat:
-				b.Put(key, float64(t))
+				b.CPut(key, float64(t), nil)
 			case parser.DString:
-				b.Put(key, string(t))
+				b.CPut(key, string(t), nil)
```
Comment (vivekmenezes) [inline sql/insert.go]: If we change these CPuts to Put it breaks the driver_test unittest. We have a problem here that is a bit mysterious.

Comment (tamird) [inline sql/insert.go]: Which CPuts did you change? Just these four?

Comment (petermattis) [inline sql/insert.go]: Why would changing these `CPuts` to `Puts` be valid? We need this to be `CPuts` so that insertion of a duplicate row is detected. Am I missing something?

Comment (vivekmenezes) [inline sql/insert.go]: Hey Pete, I happen to do it and came up with some puzzling results.
Actually what i did was incorrect, but turns out I've uncovered a bug with
us assuming that index[0] is always
a primary index. I'm working on a PR to get that fixed.

Vivek

Comment (tamird) [inline sql/insert.go]: I thought the primary index is by definition `Indexes[0]`.

Comment (vivekmenezes) [inline sql/insert.go]: By design, but the current code doesn't guarantee it. I just sent out a PR:

https://github.com/cockroachdb/cockroach/pull/1912
